cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -134,6 +142,54 @@
     this.tasks = new LinkedList<>();
     this.metricsCollectionService = metricsCollectionService;
     this.programCompletionNotifiers = Collections.emptySet();
+    this.cConf = cConf;
+    this.txRunner = transactionRunner;
+    this.store = store;
+  }
+
+  @Override
+  protected void doStartUp() throws Exception {
+    super.doStartUp();
+    AtomicBoolean completed = new AtomicBoolean();
```

Comment:
Since number of rows that we need to read might be greater than fetch size, we need to perform the read multiple times. If we see records with startTS before current time, we need to continue reading by setting completed to false. Otherwise, completed is set to true so can exit while(!completed.get()) block.

Another code sample: https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/AppMetadataStoreTest.java#L636

Reply:
hmm.. I don't quite understand here , isn't each batch fetched one by one in the single thread ?